chain/network/src/routing/graph/mod.rs

```diff
@@ -0,0 +1,315 @@
+                edges.push(e.clone());
+            }
+        }
+        edges
+    }
+
+    fn prune_old_edges(&mut self, prune_edges_older_than: time::Utc) {
+        for e in self.edges.clone().values() {
+            if e.is_edge_older_than(prune_edges_older_than) {
+                self.remove_edge(e.key());
+            }
+        }
+    }
+
+    /// If peer_id is not in memory check if it is on disk in bring it back on memory.
+    ///
+    /// Note: here an advanced example, which shows what's happening.
+    /// Let's say we have a full graph fully connected with nodes `A, B, C, D`.
+    /// Step 1 ) `A`, `B` get removed.
+    /// We store edges belonging to `A` and `B`: `<A,B>, <A,C>, <A, D>, <B, C>, <B, D>`
+    /// into component 1 let's call it `C_1`.
+    /// And mapping from `A` to `C_1`, and from `B` to `C_1`
+    ///
+    /// Note that `C`, `D` is still active.
+    ///
+    /// Step 2) 'C' gets removed.
+    /// We stored edges <C, D> into component 2 `C_2`.
+    /// And a mapping from `C` to `C_2`.
+    ///
+    /// Note that `D` is still active.
+    ///
+    /// Step 3) An active edge gets added from `D` to `A`.
+    /// We will load `C_1` and try to re-add all edges belonging to `C_1`.
+    /// We will add `<A,B>, <A,C>, <A, D>, <B, C>, <B, D>`
+    ///
+    /// Important note: `C_1` also contains an edge from `A` to `C`, though `C` was removed in `C_2`.
+    /// - 1) We will not load edges belonging to `C_2`, even though we are adding an edges from `A` to deleted `C`.
+    /// - 2) We will not delete mapping from `C` to `C_2`, because `C` doesn't belong to `C_1`.
+    /// - 3) Later, `C` will be deleted, because we will figure out it's not reachable.
+    /// New component `C_3` will be created.
+    /// And mapping from `C` to `C_2` will be overridden by mapping from `C` to `C_3`.
+    /// And therefore `C_2` component will become unreachable.
+    /// TODO(gprusak): this whole algorithm seems to be leaking stuff to storage and never cleaning up.
+    /// What is the point of it? What does it actually gives us?
+    fn load_component(&mut self, now: time::Utc, peer_id: PeerId) {
+        if peer_id == self.config.node_id || self.peer_reachable_at.contains_key(&peer_id) {
+            return;
+        }
+        let edges = match self.store.pop_component(&peer_id) {
+            Ok(edges) => edges,
+            Err(e) => {
+                tracing::warn!("self.store.pop_component({}): {}", peer_id, e);
+                return;
+            }
+        };
+        for e in edges {
+            self.update_edge(now, e);
+        }
+    }
+
+    /// Prunes peers unreachable since <unreachable_since> (and their adjacent edges)
+    /// from the in-mem graph and stores them in DB.
+    fn prune_unreachable_peers(&mut self, unreachable_since: time::Instant) {
+        // Select peers to prune.
+        let mut peers = HashSet::new();
+        for k in self.edges.keys() {
+            for peer_id in [&k.0, &k.1] {
```

Comment:
seems that we do a lot of unnecessary work here - as we might be looking at the same peer many times.

maybe add a TODO + a tracking bug to fix it.

(might be a good 'first_time_issue' for a new contributor)

Reply:
well, I'm pretty sure it is not worth optimizing:
- we go through all the edges multiple times anyway (bfs, pruning old edges, finding local edges, etc)
- we write to DB in addition to that (unless we don't find anything to prune)
- we first find all the unreachable peers and only then scan for adjacent edges, so it is not like we do a lot of work per peer_id.
- we should kill with fire the graph anyway, so it feels like a potentially wasted effort.

If you still think it is worth creating an issue for a new contributor, please let me know what kind of optimization you would like to see here.